New Vite plugin docs for local development (#20586)
cloudflare/cloudflare-docs: src/content/docs/workers/local-development/local-data.mdx

---
pcx_content_type: navigation
title: Local data
sidebar:
  order: 2
head: []
description: Working with data during local development
---

import {
	Details,
	LinkCard,
	Render,
	PackageManagers,
	FileTree,
	Aside,
} from "~/components";

> Whether you are using Wrangler or the [Cloudflare Vite plugin](https://developers.cloudflare.com/workers/vite-plugin/), your workflow for **accessing** data during local development remains the same. However, you can only [populate local resources with data](/workers/local-development/local-data/#populating-local-resources-with-data) via the Wrangler CLI.

Reviewer: ```suggestion
Whether you are using Wrangler or the [Cloudflare Vite plugin](https://developers.cloudflare.com/workers/vite-plugin/), your workflow for **accessing** data during local development remains the same. Local resources are populated with data via the Wrangler CLI.
```

### How it works

> When you run either `wrangler dev` or [`vite`](https://vite.dev/guide/cli#dev-server), [Miniflare](/workers/testing/miniflare/) automatically creates **local versions** of your resources (like [KV](/kv), [D1](/d1/), or [R2](/r2)). This means you **don’t** need to manually set up separate local instances for each service. However, newly created local resources **won’t** contain any data — you'll need to use Wrangler commands with the `--local` flag to populate them. Changes made to local resources won’t affect production data.

Reviewer: ```suggestion
When you run either `wrangler dev` or `vite dev`, [Miniflare](/workers/testing/miniflare/) automatically creates **local versions** of your resources (like [KV](/kv), [D1](/d1/), or [R2](/r2)). This means you **don’t** need to manually set up separate local instances for each service. However, newly created local resources **won’t** contain any data — you'll need to use Wrangler commands with the `--local` flag to populate them. Changes made to local resources won’t affect production data.
```

## Populating local resources with data

> When you first start developing, your local resources will be empty. You'll need to populate them with data for testing, using the Wrangler CLI. The Vite plugin doesn't provide any special functionality for populating data - it relies entirely on Wrangler CLI commands with the `--local` flag.

Reviewer: ```suggestion
When you first start developing, your local resources will be empty. You'll need to populate them with data using the Wrangler CLI.
```

### KV namespaces

<Aside type="caution" title="Syntax note">

> Since version 3.60.0, Wrangler supports the `kv ...` syntax. If you are using versions below 3.60.0, the command follows the `kv:...` syntax. Learn more about the deprecation of the `kv:...` syntax in the[Wrangler commands for KV page](/kv/reference/kv-commands/).

Reviewer 2: ```suggestion
Since version 3.60.0, Wrangler supports the `kv ...` syntax. If you are using versions below 3.60.0, the command follows the `kv:...` syntax. Learn more in the [Wrangler commands for KV page](/kv/reference/kv-commands/).
```

</Aside>

#### [Add a single key-value pair](/workers/wrangler/commands/#kv-key)

<PackageManagers
	type="exec"
	pkg="wrangler"
	args="kv key put <KEY> <VALUE> --binding=<BINDING> --local "
/>

#### [Buld upload](/workers/wrangler/commands/#kv-bulk)

<PackageManagers
	type="exec"
	pkg="wrangler"
	args="kv bulk put <FILENAME.json> --binding=<BINDING> --local"
/>

### R2 buckets

#### [Upload a file](/workers/wrangler/commands/#r2-object)

<PackageManagers
	type="exec"
	pkg="wrangler"
	args="r2 object put <BUCKET>/<KEY> --file=<PATH_TO_FILE> --local"
/>

You may also include [other metadata](/workers/wrangler/commands/#r2-object-put).

### D1 databases

> #### [Execute a SQL statement](/wrangler/commands/#d1-execute)

Reviewer 3: ```suggestion
#### [Execute a SQL statement](/workers/wrangler/commands/#d1-execute)
```

<PackageManagers
	type="exec"
	pkg="wrangler"
	args='d1 execute <DATABASE_NAME> --command="<SQL_QUERY>" --local'
/>

> #### [Execute a SQL file](/wrangler/commands/#d1-execute)

Reviewer 3: ```suggestion
#### [Execute a SQL file](/workers/wrangler/commands/#d1-execute)
```

<PackageManagers
	type="exec"
	pkg="wrangler"
	args="wrangler d1 execute <DATABASE_NAME> --file=./schema.sql --local"
/>

### Durable Objects

> For Durable Objects, unlike KV, D1, and R2, there are no CLI commands to populate them with test data. To add data to Durable Objects during local development, you must write application code that creates Durable Object instances and [calls methods on them that store state](/durable-objects/best-practices/access-durable-objects-storage/). This typically involves creating development endpoints or test routes that initialize your Durable Objects with the desired data.

Reviewer: ```suggestion
For Durable Objects, unlike KV, D1, and R2, there are no CLI commands to populate them with local data. To add data to Durable Objects during local development, you must write application code that creates Durable Object instances and [calls methods on them that store state](/durable-objects/best-practices/access-durable-objects-storage/). This typically involves creating development endpoints or test routes that initialize your Durable Objects with the desired data.
```

	args="kv key put test 12345 --binding MY_KV_NAMESPACE --local --persist-to worker-local"
/>

This command:

- Sets the KV key `test` to `12345` in the binding `MY_KV_NAMESPACE` (defined in your [Wrangler configuration file](/workers/wrangler/configuration/)).
- Uses `--persist-to worker-local` to ensure the data is created in the **worker-local** directory instead of the default `.wrangler/state`.
- Adds the `--local` flag, indicating you want to modify local data.

If `--persist-to` is not specified, Wrangler defaults to using `.wrangler/state` for local data.

</Details>

#### Using the Cloudflare Vite plugin

To customize where the Vite plugin stores local data, configure the [`persistState` option](/workers/vite-plugin/reference/api/#interface-pluginconfig) in your Vite config file:

```js title="vite.config.js"
import { defineConfig } from "vite";
import { cloudflare } from "@cloudflare/vite-plugin";

export default defineConfig({
	plugins: [
		cloudflare({
			persistState: "./my-custom-directory",
		}),
	],
});
```

#### Sharing state between tools

If you want Wrangler and the Vite plugin to share the same state, configure them to use the same persistence path.

> ## Customize `wrangler dev`
> 
> You can further customize the behavior of `wrangler dev` (for instance, by changing ports or skipping Wrangler's build steps). Refer to [the `wrangler dev` documentation](/workers/wrangler/commands/#dev) for available configuration options.
> 
> ## Customize the Cloudflare Vite plugin
> 
> You can also [customize the Vite plugin](/workers/vite-plugin/reference/api/), such as changing ports or calling other Workers.
> 
> ## Debugging
> 
> - [DevTools](/workers/observability/dev-tools) - Guides to using DevTools to debug your Worker locally.
> - [Debugging with the Vite plugin](/workers/vite-plugin/reference/debugging/)

Reviewer: These sections don't seem relevant to 'Local data'.

```suggestion
```
